### PR TITLE
Implement `log2` and `log` functions in math.tact

### LINKED ISSUE #12: stdlib: logarithm calculation
stdlib need a gas-efficient logarithm calculation

### PR DESCRIPTION
**Pull request by Gusarich**
closes #12

### CODE REVIEW

File: src/generator/writers/writeStdlib.ts [hunk 1]
```diff
@@ -1223,4 +1223,20 @@ export function writeStdlib(ctx: WriterContext) {
...
+        ctx.context('stdlib');
+        ctx.asm(`asm "UBITSIZE DEC"`);
+    });
+
+    ctx.fun(`__tact_log`, () => {
+        ctx.signature(`int __tact_log(int num, int base)`);
+        ctx.context('stdlib');
+        ctx.body(() => {
+            ctx.write(`
+                return __tact_log2(num) / __tact_log2(base);
```
Comment (anton-trunov) [inline src/generator/writers/writeStdlib.ts]: This formula does not work for integer logarithms with an arbitrary base. We basically need `floor(log(num, base))`, which, for example, for `log3(2^32 - 1)` gives us 20 as the result, not 32, which this formula provides.

File: src/test/feature-math.spec.ts [hunk 2]
```diff
@@ -335,5 +338,21 @@ describe('feature-math', () => {
...
+            expect(await contract.getLog2(num)).toBe(
+                BigInt(Math.floor(Math.log2(Number(num))))
+            );
+        }
+
+        for (let num = 1n; num <= 10n; num++) {
+            for (let base = 2n; base <= 10; base++) {
+                const a = Math.floor(Math.log2(Number(num)));
+                const b = Math.floor(Math.log2(Number(base)));
+                const c = BigInt(Math.floor(a / b));
```
Comment (anton-trunov) [inline src/test/feature-math.spec.ts]: we want `c` to be something like `floor(log2(num) / log2(base))`, just like in the comment above

File: src/test/feature-math.spec.ts [hunk 3]
```diff
@@ -335,5 +338,21 @@ describe('feature-math', () => {
         expect(await contract.getCompare28(dictA, dictB)).toBe(true);
         expect(await contract.getCompare28(dictB, dictA)).toBe(true);
         expect(await contract.getCompare28(dictA, dictA)).toBe(false);
+
+        // Test advanced math operations
+        for (let num = 1n; num <= 100n; num++) {
```
Comment (anton-trunov) [inline src/test/feature-math.spec.ts:343]: Please add more tests for corner cases:
- `num` is 0
- `num` is near the maximum integer value, i.e. `[maxint - 99; maxint]`
- `num` is in the middle of the range of the positive integers, i.e. `[maxint/2 - 50, maxint/2 + 50]`

File: src/test/feature-math.spec.ts [hunk 4]
```diff
@@ -335,5 +338,21 @@ describe('feature-math', () => {
...
+            expect(await contract.getLog2(num)).toBe(
+                BigInt(Math.floor(Math.log2(Number(num))))
+            );
+        }
+
+        for (let num = 1n; num <= 10n; num++) {
```
Comment (anton-trunov) [inline src/test/feature-math.spec.ts:349]: Please add more tests for corner cases:
- `num` is 0
- `num` is near the maximum integer value, i.e. `[maxint - 99; maxint]`
- `num` is in the middle of the range of the positive integers, i.e. `[maxint/2 - 50, maxint/2 + 50]`

**Review by anton-trunov — CHANGES_REQUESTED** [review]

File: src/generator/writers/writeStdlib.ts [hunk 5]
```diff
@@ -1235,7 +1235,15 @@ export function writeStdlib(ctx: WriterContext) {
         ctx.context('stdlib');
         ctx.body(() => {
             ctx.write(`
-                return __tact_log2(num) / __tact_log2(base);
+                if (num < base) {
+                    return 0;
+                }
+                int result = 0;
+                while (num >= base) {
+                    num = num / base;
```
Comment (anton-trunov) [inline src/generator/writers/writeStdlib.ts]:
```suggestion
                    num /= base;
```

or would it be less efficient in FunC?

File: src/test/feature-math.spec.ts [hunk 6]
```diff
@@ -348,10 +348,51 @@ describe('feature-math', () => {
...
+        const maxint = 2n ** 256n - 1n;
+
+        function bigIntLogBase(num: bigint, base: bigint) {
+            let result = 0n;
+            while (num >= base) {
+                num = num / base;
+                result += 1n;
+            }
+            return result;
+        }
```
Comment (anton-trunov) [inline src/test/feature-math.spec.ts]: This should be basically something like `Math.floor(Math.log2(Number(num)) / Math.log2(Number(base)))` as above. Otherwise the testing code just repeats the code being tested.

**Review by anton-trunov — CHANGES_REQUESTED** [review]

**Review by anton-trunov — APPROVED** [review]
Cool solution with `toString(base).length - 1` for tests!
